server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -169,28 +173,32 @@ public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
     /**
      * A base class for health metadata cluster state update tasks.
      */
-    abstract static class UpsertHealthMetadataTask implements ClusterStateTaskListener {
+    interface UpsertHealthMetadataTask extends ClusterStateTaskListener {
 
         @Override
-        public void onFailure(@Nullable Exception e) {
+        default void onFailure(@Nullable Exception e) {
             logger.log(
                 MasterService.isPublishFailureException(e) ? Level.DEBUG : Level.WARN,
                 () -> "failure during health metadata update",
                 e
             );
         }
 
-        ClusterState execute(ClusterState currentState) {
+        default ClusterState execute(ClusterState currentState) {
+            if (currentState.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
```

Comment:
we cannot add this check in the first line of `updateOnClusterStateChange` because it'll block the logic to mark this node as master. We could add this check right after line 146, but I preferred to put it here because it's the moment before publishing potential clusterstate changes, and it's shared among all the task received (`resetHealthMetadata` and updates). _Not sure whether it's the right place, though_. I can guess we can still add load to the master node because we're enqueuing tasks, but I don't know the internals to be fully sure.

Reply:
> it'll block the logic to mark this node as master

That's correct. But in this case we would just assume we are not the master in the meantime, and not do any health related things. I think that's ok isn't it? 

I think it's ok (preferable even) to **not** decide we are the master until the cluster state is fully recovered.

Reply:
nevermind, my guess was correct